src/kafka-util/src/admin.rs

```diff
@@ -0,0 +1,142 @@
+// Copyright Materialize, Inc. All rights reserved.
+//
+// Use of this software is governed by the Business Source License
+// included in the LICENSE file.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0.
+
+//! Helpers for working with Kafka's admin API.
+
+use std::error::Error;
+use std::fmt;
+use std::iter;
+use std::time::Duration;
+
+use rdkafka::admin::{AdminClient, AdminOptions, NewTopic};
+use rdkafka::client::ClientContext;
+use rdkafka::error::{KafkaError, RDKafkaError};
+
+use ore::collections::CollectionExt;
+use ore::retry;
+
+/// Creates a Kafka topic and waits for it to be reported in the broker
+/// metadata.
+///
+/// This function is a wrapper around [`AdminClient::create_topics`] that
+/// attempts to ensure the topic creation has propagated throughout the Kafka
+/// cluster before returning. Kafka topic creation is asynchronous, so
+/// attempting to consume from or produce to a topic immediately after its
+/// creation can result in "unknown topic" errors.
+///
+/// This function does not return successfully unless it can find the metadata
+/// for the newly-created topic in a call to [`Client::fetch_metadata`] and
+/// verify that the metadata reports the topic has the number of partitions
+/// requested in `new_topic`. Empirically, this seems to be the condition that
+/// guarantees that future attempts to consume from or produce to the topic will
+/// succeed.
+pub async fn create_topic<'a, C>(
+    client: &'a AdminClient<C>,
+    admin_opts: &AdminOptions,
+    new_topic: &'a NewTopic<'a>,
+) -> Result<(), CreateTopicError>
+where
+    C: ClientContext,
+{
+    let res = client
+        .create_topics(iter::once(new_topic), &admin_opts)
+        .await?;
+    if res.len() != 1 {
+        return Err(CreateTopicError::InvariantViolation(format!(
+            "kafka topic creation returned {} results, but exactly one result was expected",
+            res.len()
+        )));
+    }
+    match res.into_element() {
+        Ok(_) | Err((_, RDKafkaError::TopicAlreadyExists)) => Ok(()),
+        Err((_, e)) => Err(CreateTopicError::Kafka(KafkaError::AdminOp(e))),
+    }?;
+
+    // Topic creation is asynchronous, and if we don't wait for it to
```

Comment:
I think we need to remove the bits in this comment that talk about single partition-ness because this function doesnt actually care about that afaict